### PR TITLE
Export Hooks: Glue Catalog exporter

### LINKED ISSUE #6575: Export Hooks: Glue Catalog exporter
- Based on design #6461 
- Depends #6574

# D.O.D 

-  lakeFS users can use lakeFS as storage and then query the data in Glue Catalog and clients (I.e Athena, Trino etc) 
- Users can specify a table spec in Hive format in `_lakefs_tables` directory + Action events upon which they want the table to be updated. 
- see [glue_exporter.lua](https://github.com/treeverse/lakeFS/issues/6461) in the design.

# Implementation details 

- Extend Lua package `lakefs/catalog_export`, reuse the logic that exports Symlink to S3,  add updates  to glue with schema. 
- Require packages: Symlink Exporter dependencies, glue client / metastore primitives implemented

### PR DESCRIPTION
**Pull request by Isan-Rivkin**
Closes #6575 

# TODO - will create additional PR and merge into current: 
- Add system test with Glue / AWS 
- Add documentation 

# Usage 

### Pre-requisite  
  
Write some table data to lakeFS either `parquet` or `CSV` format, snippet of parquet added:

```python
data =[("James","","Smith","36636","M",3000),
              ("Michael","Rose","","40288","M",4000),
              ("Robert","","Williams","42114","M",4000),
              ("Maria","Anne","Jones","39192","F",4000),
              ("Jen","Mary","Brown","","F",-1)]

columns=["firstname","middlename","lastname","dob","gender","salary"]

df=spark.createDataFrame(data,columns)
```

 
### Define table in `_lakefs_tables/employees.yaml` 

```yaml
name: employees
type: hive
path: <path/to/parquet/files>
partition_columns: ['gender', 'salary']
schema:
  type: struct
  fields:
    - name: gender
      type: string
      nullable: false
      metadata: {}
    - name: salary
      type: integer
      nullable: false
      metadata: {}
    - name: firstname
      type: string
      nullable: false
      metadata: {}
    - name: middlename
      type: string
      nullable: false
      metadata: {}
    - name: lastname
      type: string
      nullable: false
      metadata: {}
    - name: dob
      type: string
      nullable: false
      metadata: {}

```

### Create `_lakefs_action/employee_exporter.yaml`

```yaml
name: glue catalog exporter
on:
  post-create-branch:
    branches: ["main*"]
  post-commit:
    branches: ["main*"]
hooks:
  # symlink_exporter placegolder - can be separated to create symlink file
  - id: symlink_exporter
  - id: glue_exporter
    type: lua
    properties:
      script_path: scripts/glue_export.lua
      args:
        # Export configuration
        aws:
          aws_access_key_id: ''
          aws_secret_access_key: ''
          aws_region: ''
        table_source: '_lakefs_tables/employees.yaml'
        catalog:
          db_name: "<glue-database-name>"
          # csv input example, can be passed from here or from script file
          # table_input:
          #   StorageDescriptor: 
          #     InputFormat: "org.apache.hadoop.hive.ql.io.SymlinkTextInputFormat"
          #     # InputFormat: "org.apache.hadoop.mapred.TextInputFormat"
          #     OutputFormat: "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
          #     SerdeInfo:
          #         SerializationLibrary: "org.apache.hadoop.hive.serde2.OpenCSVSerde"
          #         Parameters:
          #             separatorChar: ","
          #   Parameters: 
          #     classification: "csv"
```

### Create export hook in `scripts/glue_export.lua`

```lua
local aws = require("aws")
local exporter = require("lakefs/catalogexport/glue_exporter")

-- config 
local table_path = args.table_source
local db = args.catalog.db_name
local options = {
    debug=true,
}

-- init glue client 
local glue = aws.glue_client(args.aws.aws_access_key_id, args.aws.aws_secret_access_key, args.aws.aws_region)

-- configure parquet
local table_input = {
    StorageDescriptor={
        InputFormat= "org.apache.hadoop.hive.ql.io.SymlinkTextInputFormat",
        OutputFormat= "org.apache.hadoop.hive.ql.io.IgnoreKeyTextOutputFormat",
        SerdeInfo={
            SerializationLibrary= "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
        },
    },
    Parameters={
        classification= "parquet",
        "parquet.compression"= "SNAPPY",
    }
}

-- export glue table 
local result = exporter.export_glue(glue, db, table_path, table_input, action, options)
print("success exporting table to glue location", result.create_table_input.StorageDescriptor.Location)
```

### CODE REVIEW

**Comment by Isan-Rivkin** [comment]
@nopcoder please notice that PR description added usage (TODO's will be done after this PR code approved)

File: pkg/actions/lua/storage/aws/glue.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		// Unmarshal the JSON to a map.
+		var itemMap map[string]interface{}
+		err = json.Unmarshal(jsonBytes, &itemMap)
+		if err != nil {
+			lua.Errorf(l, err.Error())
+			panic("unreachable")
+		}
+		retArgs := util.DeepPush(l, itemMap)
+		l.PushBoolean(true)
+		return retArgs + 1
```
Comment (nopcoder) [inline pkg/actions/lua/storage/aws/glue.go:222]: Strange that DeepPush needs to return a number - it expected to already return 1.
Same for this function `getTable` return - we must return 2 as we like the boolean to be on position 2.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/actions/lua/storage/aws/glue.go (hunk 1 above)
Comment (Isan-Rivkin) [reply pkg/actions/lua/storage/aws/glue.go:222]: Good catch, changed 👍

File: pkg/actions/lua/storage/aws/glue.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+		database := lua.CheckString(l, 1)
+		tableInputJSON := lua.CheckString(l, 2)
+
+		// parse table input JSON
+		var tableInput types.TableInput
+		err := json.Unmarshal([]byte(tableInputJSON), &tableInput)
+		if err != nil {
+			lua.Errorf(l, err.Error())
+			panic("unreachable")
+		}
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/storage/aws/glue.go:115]: 👍

File: pkg/actions/lua/storage/aws/glue.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	"get_table":    getTable,
+	"create_table": createTable,
+	"update_table": updateTable,
+	"delete_table": deleteTable,
+}
+
+func (c *GlueClient) client() *glue.Client {
+	cfg, err := config.LoadDefaultConfig(c.ctx,
+		config.WithRegion(c.Region),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(c.AccessKeyID, c.SecretAccessKey, "")),
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/storage/aws/glue.go:67]: Verified, works :)

File: pkg/actions/lua/lakefs/catalogexport/internal.lua [hunk 4]
```diff
@@ -1,6 +1,18 @@
 local url = require("net/url")
 local DEFAULT_SHORT_DIGEST_LEN=6
 
+local function clone_table(original_table)
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/internal.lua:4]: woha took me a while to compile that in my mind, but better for sure - updated thanks.

File: pkg/actions/lua/lakefs/catalogexport/hive.lua [hunk 5]
```diff
@@ -68,6 +68,9 @@ local function extract_partition_pager(client, repo_id, commit_id, base_path, pa
                 })
                 -- remove entry only if its part of the current partition
                 table.remove(page, 1)
+            else
+                -- remove hidden files from the entry set
+                table.remove(page, 1)
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/hive.lua:73]: 👍

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    local table_input = opts.override_create_table_input or build_glue_create_table_input_from_hive(descriptor, create_table_table_input, action_info, opts)
+    -- create table
+    local json_input = json.marshal(table_input)
+    if opts.debug then
+        print("Creating Glue Table - input: ", json_input)
+    end
+    glue.create_table(db, json_input)
+    return {
+        create_table_input=table_input
+    }
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua:117]: To leave room for other variables and make it a "result" without breaking function signature. 

For example, in a more complex scenario where export will do version updates to the same table we will want the new latest version things like that.

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    local commit_id = action_info.commit_id
+    -- get table desctiptor
+    local descriptor = extractor.get_table_descriptor(lakefs, repo_id, commit_id, table_src_path)
+
+    -- build table creation input for glue
+    local table_input = opts.override_create_table_input or build_glue_create_table_input_from_hive(descriptor, create_table_table_input, action_info, opts)
+    -- create table
+    local json_input = json.marshal(table_input)
+    if opts.debug then
+        print("Creating Glue Table - input: ", json_input)
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua:112]: 👍

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    local symlink_location = get_table_location(base_prefix, descriptor, action_info)
+
+    -- convert hive cols/partitions to glue
+    local partitions = hive_partitions_to_glue_input(descriptor)
+    local cols = hive_columns_to_glue_input(descriptor)
+    return build_glue_create_table_input(base_input, descriptor, symlink_location, cols, partitions, action_info, opts)
+end
+
+-- create table in glue
+local function export_glue(glue, db, table_src_path,create_table_table_input, action_info, options)
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua:100]: 👍

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    return pathlib.join("/", export_base_uri, descriptor.name)
+end
+
+-- create a standard AWS Glue table input (i.e not Apache Iceberg), add input values to base input and configure the rest
+local function build_glue_create_table_input(base_input, descriptor, symlink_location, columns, partitions, action_info,
+    options)
+    local input = utils.clone_table(base_input)
+    local opts = options or {}
+    local repo_id = action_info.repository_id
+    local commit_id = action_info.commit_id
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua:78]: 👍

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+-- default location value (e.g root location of either partitions or flat symlink.txt file)
+local function get_table_location(storage_base_prefix, descriptor, action_info)
+    local commit_id = action_info.commit_id
+    local export_base_uri = sym_exporter.get_storage_uri_prefix(storage_base_prefix, commit_id, action_info)
+    return pathlib.join("/", export_base_uri, descriptor.name)
+end
+
+-- create a standard AWS Glue table input (i.e not Apache Iceberg), add input values to base input and configure the rest
+local function build_glue_create_table_input(base_input, descriptor, symlink_location, columns, partitions, action_info,
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua:73]: Yes, I see usage incase user want's to use `opts.override_create_table_input` and change certain things in `build_glue_create_table_input` output. 
Actually I added now export for this function as well outside of the module.

**Comment by Isan-Rivkin** [comment]
@nopcoder Thanks! Please re-review

File: pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/actions/lua/lakefs/catalogexport/glue_exporter.lua]: changed to table_input

**Review by nopcoder — APPROVED** [review]
LGTM!

**Pull request merged**